rasa/core/trackers.py

```diff
@@ -420,11 +432,67 @@ def undo_till_previous(event_type, done_events):
                 # the `action_listen` action right before it).
                 undo_till_previous(UserUttered, applied_events)
                 undo_till_previous(ActionExecuted, applied_events)
+            elif (
+                isinstance(event, ActionExecuted)
+                and event.action_name in form_names
+                and not self._upcoming_rejection(idx)
+                and not self._first_loop_execution_or_previous_rejection(
+                    event.action_name, applied_events
+                )
+            ):
+                self.undo_till_previous_loop_execution(
+                    event.action_name, applied_events
+                )
+            elif isinstance(event, ActionExecutionRejected):
+                # Undo previously executed action
+                undo_till_previous(ActionExecuted, applied_events)
+                applied_events.append(event)
             else:
                 applied_events.append(event)
 
         return applied_events
 
+    def _upcoming_rejection(self, index: int) -> bool:
+        return next(
+            (
+                isinstance(e, ActionExecutionRejected)
+                for e in list(self.events)[index + 1 :]
+                if isinstance(e, (ActionExecutionRejected, ActionExecuted))
```

Comment:
will it terminate `for` after first `if`, or generate the whole list? we care only for the first `if`

Reply:
it's a generator so it'll go for the first `if`